File: apps/windows/Finance/GeminiReceiptParser.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Andivum_Windows.Finance;

public sealed record GeminiRequest(string ApiKeyHeader, string Body);

public static class GeminiReceiptParser
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true,
    };

    public static GeminiRequest BuildRequest(
        byte[] document,
        string mimeType,
        IReadOnlyCollection<FinanceCategory> categories,
        string apiKey,
        string model = "gemini-2.5-flash")
    {
        if (document.Length == 0)
        {
            throw new ArgumentException("Document is empty.", nameof(document));
        }
        if (string.IsNullOrWhiteSpace(apiKey))
        {
            throw new ArgumentException("Gemini API key is required.", nameof(apiKey));
        }
        if (string.IsNullOrWhiteSpace(mimeType))
        {
            throw new ArgumentException("Document MIME type is required.", nameof(mimeType));
        }

        var allowedCategories = categories
            .Where(category => !string.IsNullOrWhiteSpace(category.Slug))
            .Select(category => new
            {
                category.Slug,
                category.Name,
                category.Type,
            })
            .ToArray();
        var schema = new
        {
            type = "OBJECT",
            properties = new
            {
                title = new { type = "STRING" },
                type = new { type = "STRING", @enum = new[] { "income", "expense" } },
                occurred_on = new { type = "STRING" },
                currency = new { type = "STRING" },
                total_minor = new { type = "INTEGER" },
                items = new
                {
                    type = "ARRAY",
                    items = new
                    {
                        type = "OBJECT",
                        properties = new
                        {
                            name = new { type = "STRING" },
                            quantity = new { type = "NUMBER" },
                            unit_price_minor = new { type = "INTEGER" },
                            line_total_minor = new { type = "INTEGER" },
                            category_slug = new { type = "STRING" },
                        },
                        required = new[] { "name", "quantity", "unit_price_minor", "line_total_minor", "category_slug" },
                    },
                },
            },
            required = new[] { "title", "type", "occurred_on", "currency", "total_minor", "items" },
        };
        var prompt = """
            Extract one personal-finance receipt or invoice into the JSON schema.
            Return only JSON. Never invent an amount or date. If a value is not visible,
            use an empty string and the UI will ask the user. total_minor and every money
            field are integer minor units in the currency. type must be income or expense.
            category_slug must be exactly one of the allowed categories below; otherwise
            use other.expense or income.other. The result is a draft, not an authorization.
            Allowed categories:
            """ + JsonSerializer.Serialize(allowedCategories, JsonOptions);
        var body = new
        {
            contents = new[]
            {
                new
                {
                    parts = new object[]
                    {
                        new { text = prompt },
                        new
                        {
                            inline_data = new
                            {
                                mime_type = mimeType,
                                data = Convert.ToBase64String(document),
                            },
                        },
                    },
                },
            },
            generationConfig = new
            {
                responseMimeType = "application/json",
                responseSchema = schema,
                temperature = 0.1,
            },
        };
        return new GeminiRequest(
            apiKey,
            JsonSerializer.Serialize(body, JsonOptions));
    }

    public static async Task<FinanceDraft> ParseAsync(
        HttpClient httpClient,
        byte[] document,
        string mimeType,
        IReadOnlyCollection<FinanceCategory> categories,
        string apiKey,
        string model = "gemini-2.5-flash",
        CancellationToken cancellationToken = default)
    {
        var request = BuildRequest(document, mimeType, categories, apiKey, model);
        using var message = new HttpRequestMessage(
            HttpMethod.Post,
            $"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent")
        {
            Content = new StringContent(request.Body, Encoding.UTF8, "application/json"),
        };
        message.Headers.Add("x-goog-api-key", request.ApiKeyHeader);
        using var response = await httpClient.SendAsync(message, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                "Gemini could not analyze the document. Check the provider settings or try again later.");
        }

        return ParseResponse(
            await response.Content.ReadAsStringAsync(cancellationToken),
            categories);
    }

    public static FinanceDraft ParseResponse(
        string response,
        IReadOnlyCollection<FinanceCategory> categories)
    {
        using var document = JsonDocument.Parse(response);
        var candidateText = document.RootElement
            .GetProperty("candidates")[0]
            .GetProperty("content")
            .GetProperty("parts")[0]
            .GetProperty("text")
            .GetString();
        if (string.IsNullOrWhiteSpace(candidateText))
        {
            throw new InvalidOperationException("Gemini returned an empty finance draft.");
        }

        var json = candidateText.Trim();
        if (json.StartsWith("```", StringComparison.Ordinal))
        {
            json = json.Trim('`', ' ', 'j', 's', 'o', 'n', '\r', '\n');
        }
        var draft = JsonSerializer.Deserialize<GeminiDraft>(json, JsonOptions)
            ?? throw new InvalidOperationException("Gemini returned an invalid finance draft.");
        var expectedType = (draft.Type ?? string.Empty).ToLowerInvariant() switch
        {
            "income" => "income",
            "expense" => "expense",
            _ => throw new InvalidOperationException("Gemini returned an unsupported transaction type."),
        };
        var available = categories
            .Where(category => category.Type.Equals(expectedType, StringComparison.OrdinalIgnoreCase))
            .Select(category => category.Slug)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);
        var items = draft.Items?.Select(item =>
        {
            if (string.IsNullOrWhiteSpace(item.Name) ||
                item.Quantity <= 0 ||
                item.LineTotalMinor < 0 ||
                !available.Contains(item.CategorySlug ?? string.Empty))
            {
                throw new InvalidOperationException("Gemini returned an invalid finance item.");
            }
            return new FinanceDraftItem(
                item.Name,
                item.Quantity,
                item.LineTotalMinor,
                item.CategorySlug!);
        }).ToArray() ?? throw new InvalidOperationException("Gemini returned no finance items.");

        return new FinanceDraft(
            expectedType.Equals("income", StringComparison.Ordinal)
                ? FinanceTransactionType.Income
                : FinanceTransactionType.Expense,
            string.IsNullOrWhiteSpace(draft.Title) ? "Без названия" : draft.Title.Trim(),
            draft.OccurredOn ?? string.Empty,
            (draft.Currency ?? string.Empty).ToUpperInvariant(),
            draft.TotalMinor,
            items);
    }

    private sealed record GeminiDraft(
        [property: JsonPropertyName("title")] string? Title,
        [property: JsonPropertyName("type")] string? Type,
        [property: JsonPropertyName("occurred_on")] string? OccurredOn,
        [property: JsonPropertyName("currency")] string? Currency,
        [property: JsonPropertyName("total_minor")] long TotalMinor,
        [property: JsonPropertyName("items")] IReadOnlyList<GeminiItem>? Items);

    private sealed record GeminiItem(
        [property: JsonPropertyName("name")] string? Name,
        [property: JsonPropertyName("quantity")] decimal Quantity,
        [property: JsonPropertyName("line_total_minor")] long LineTotalMinor,
        [property: JsonPropertyName("category_slug")] string? CategorySlug);
}

File: apps/android/app/src/main/java/io/github/andivalsei/andivum/GeminiReceiptParser.kt
package io.github.andivalsei.andivum

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONArray
import org.json.JSONObject

data class GeminiRequest(
    val apiKeyHeader: String,
    val body: String,
)

object GeminiReceiptParser {
    fun buildRequest(
        bytes: ByteArray,
        mimeType: String,
        categories: Collection<FinanceCategory>,
        apiKey: String,
        model: String = "gemini-2.5-flash",
    ): GeminiRequest {
        require(bytes.isNotEmpty()) { "Document is empty." }
        require(mimeType.isNotBlank()) { "Document MIME type is required." }
        require(apiKey.isNotBlank()) { "Gemini API key is required." }

        val allowed = JSONArray()
        categories.filter { it.slug.isNotBlank() }.forEach { category ->
            allowed.put(
                JSONObject()
                    .put("slug", category.slug)
                    .put("name", category.name)
                    .put("type", category.type),
            )
        }
        val schema = JSONObject()
            .put("type", "OBJECT")
            .put(
                "properties",
                JSONObject()
                    .put("title", JSONObject().put("type", "STRING"))
                    .put("type", JSONObject().put("type", "STRING"))
                    .put("occurred_on", JSONObject().put("type", "STRING"))
                    .put("currency", JSONObject().put("type", "STRING"))
                    .put("total_minor", JSONObject().put("type", "INTEGER"))
                    .put(
                        "items",
                        JSONObject()
                            .put("type", "ARRAY")
                            .put(
                                "items",
                                JSONObject()
                                    .put("type", "OBJECT")
                                    .put(
                                        "properties",
                                        JSONObject()
                                            .put("name", JSONObject().put("type", "STRING"))
                                            .put("quantity", JSONObject().put("type", "NUMBER"))
                                            .put("unit_price_minor", JSONObject().put("type", "INTEGER"))
                                            .put("line_total_minor", JSONObject().put("type", "INTEGER"))
                                            .put("category_slug", JSONObject().put("type", "STRING")),
                                    )
                                    .put(
                                        "required",
                                        JSONArray(listOf("name", "quantity", "unit_price_minor", "line_total_minor", "category_slug")),
                                    ),
                            ),
                    ),
            )
            .put("required", JSONArray(listOf("title", "type", "occurred_on", "currency", "total_minor", "items")))
        val prompt = """
            Extract one personal-finance receipt or invoice into the JSON schema.
            Return only JSON. Never invent an amount or date. Missing values must be empty.
            Money fields are integer minor units. type is income or expense. category_slug
            must be one of the allowed slugs below. This is only an editable draft.
            Allowed categories: $allowed
        """.trimIndent()
        val body = JSONObject()
            .put(
                "contents",
                JSONArray().put(
                    JSONObject().put(
                        "parts",
                        JSONArray()
                            .put(JSONObject().put("text", prompt))
                            .put(
                                JSONObject().put(
                                    "inline_data",
                                    JSONObject()
                                        .put("mime_type", mimeType)
                                        .put("data", android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP)),
                                ),
                            ),
                    ),
                ),
            )
            .put(
                "generationConfig",
                JSONObject()
                    .put("responseMimeType", "application/json")
                    .put("responseSchema", schema)
                    .put("temperature", 0.1),
            )
        return GeminiRequest(apiKey, body.toString())
    }

    fun parse(
        bytes: ByteArray,
        mimeType: String,
        categories: Collection<FinanceCategory>,
        apiKey: String,
        model: String = "gemini-2.5-flash",
        openConnection: (URL) -> HttpURLConnection = { url -> url.openConnection() as HttpURLConnection },
    ): FinanceDraft {
        val request = buildRequest(bytes, mimeType, categories, apiKey, model)
        val connection = openConnection(
            URL("https://generativelanguage.googleapis.com/v1beta/models/$model:generateContent"),
        ).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("x-goog-api-key", request.apiKeyHeader)
            connectTimeout = 20_000
            readTimeout = 60_000
            doOutput = true
        }
        try {
            connection.outputStream.use { it.write(request.body.toByteArray(Charsets.UTF_8)) }
            val statusCode = connection.responseCode
            val stream = if (statusCode in 200..299) connection.inputStream else connection.errorStream
            val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (statusCode !in 200..299) {
                throw IllegalStateException("Gemini could not analyze the document.")
            }
            return parseResponse(response, categories)
        } finally {
            connection.disconnect()
        }
    }

    fun parseResponse(response: String, categories: Collection<FinanceCategory>): FinanceDraft {
        val candidate = JSONObject(response)
            .getJSONArray("candidates")
            .getJSONObject(0)
            .getJSONObject("content")
            .getJSONArray("parts")
            .getJSONObject(0)
            .getString("text")
            .trim()
        val jsonText = candidate.removeCodeFence()
        val json = JSONObject(jsonText)
        val type = json.optString("type").lowercase()
        require(type == "income" || type == "expense") { "Gemini returned an unsupported transaction type." }
        val allowed = categories
            .filter { it.type.equals(type, ignoreCase = true) }
            .map { it.slug }
            .toSet()
        val itemRows = json.optJSONArray("items")
            ?: throw IllegalStateException("Gemini returned no finance items.")
        val items = buildList(itemRows.length()) {
            repeat(itemRows.length()) { index ->
                val item = itemRows.getJSONObject(index)
                val categorySlug = item.optString("category_slug")
                require(
                    item.optString("name").isNotBlank() &&
                        item.optDouble("quantity", 0.0) > 0.0 &&
                        item.optLong("line_total_minor", -1L) >= 0L &&
                        categorySlug in allowed,
                ) { "Gemini returned an invalid finance item." }
                add(
                    FinanceDraftItem(
                        name = item.getString("name"),
                        quantity = item.getDouble("quantity"),
                        lineTotalMinor = item.getLong("line_total_minor"),
                        categorySlug = categorySlug,
                    ),
                )
            }
        }
        val totalMinor = json.getLong("total_minor")
        require(items.sumOf { it.lineTotalMinor } == totalMinor) {
            "Gemini item totals do not equal the transaction total."
        }
        return FinanceDraft(
            type = if (type == "income") FinanceTransactionType.INCOME else FinanceTransactionType.EXPENSE,
            title = json.optString("title").ifBlank { "Без названия" },
            occurredOn = json.optString("occurred_on"),
            currency = json.optString("currency").uppercase(),
            totalMinor = totalMinor,
            items = items,
        )
    }

    private fun String.removeCodeFence(): String =
        if (startsWith("```")) trim('`', ' ', 'j', 's', 'o', 'n', '\r', '\n') else this
}
